diff --git a/src/sdl2_wrapper.cpp b/src/sdl2_wrapper.cpp
index f9a743c..8966442 100644
--- a/src/sdl2_wrapper.cpp
+++ b/src/sdl2_wrapper.cpp
@@ -216,6 +216,9 @@ CAMLprim value resdl_SDL_GetNativeWindow(value vWin) {
   case SDL_SYSWM_COCOA:
     pNativeWindow = (void *)wmInfo.info.cocoa.window;
     break;
+#elif __ANDROID__
+  case SDL_SYSWM_ANDROID:
+    pNativeWindow = (void *)wmInfo.info.android.window;
 #else
   case SDL_SYSWM_X11:
     pNativeWindow = (void *)wmInfo.info.x11.window;
@@ -1159,7 +1162,11 @@ CAMLprim value resdl_SDL_CreateWindow(value vWidth, value vHeight,
   // Attributes pulled from:
   // https://github.com/google/skia/blob/master/example/SkiaSDLExample.cpp
   static const int kStencilBits = 8; // Skia needs 8 stencil bits
+#if defined(TARGET_OS_IPHONE) || defined(__ANDROID__)
+  SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_ES);
+#else
   SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);
+#endif
 #ifdef WIN32
   SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
   SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 0);
