diff --git a/src/sdl2.re b/src/sdl2.re
index 885ba65..384b01b 100644
--- a/src/sdl2.re
+++ b/src/sdl2.re
@@ -272,6 +272,9 @@ module Scancode = {
 
 module Keycode = {
   type t = int;
+  // as some keycodes are bigger than 2^30 but still less than 2^31
+  // so by using the negative numbers we can fit them in a normal int
+  let t = Int32.to_int;
 
   external getName: t => string = "resdl_SDL_GetKeyName";
 
@@ -284,62 +287,62 @@ module Keycode = {
 
   // Incrementally add these as needed from:
   // https://wiki.libsdl.org/SDLKeycodeLookup
-  let unknown = 0;
-  let backspace = 8;
-  let return = 13;
-
-  let escape = 27;
-
-  let space = 32;
-
-  let left_paren = 40;
-  let right_paren = 41;
-
-  let asterisk = 42;
-  let plus = 43;
-  let minus = 45;
-  let period = 46;
-  let slash = 47;
-  let caret = 94;
-
-  let equals = 61;
-
-  let digit0 = 48;
-  let digit1 = 49;
-  let digit2 = 50;
-  let digit3 = 51;
-  let digit4 = 52;
-  let digit5 = 53;
-  let digit6 = 54;
-  let digit7 = 55;
-  let digit8 = 56;
-  let digit9 = 57;
-
-  let pad_divide = 1073741908;
-  let pad_multiply = 1073741909;
-  let pad_minus = 1073741910;
-  let pad_plus = 1073741911;
-  let pad_period = 1073741923;
-
-  let pad_equals = 1073741927;
-
-  let p_digit1 = 1073741913;
-  let p_digit2 = 1073741914;
-  let p_digit3 = 1073741915;
-  let p_digit4 = 1073741916;
-  let p_digit5 = 1073741917;
-  let p_digit6 = 1073741918;
-  let p_digit7 = 1073741919;
-  let p_digit8 = 1073741920;
-  let p_digit9 = 1073741921;
-  let p_digit0 = 1073741922;
-
-  let c = 99;
-
-  let delete = 127;
-
-  let right = 1073741903;
-  let left = 1073741904;
+  let unknown = t(0l);
+  let backspace = t(8l);
+  let return = t(13l);
+
+  let escape = t(27l);
+
+  let space = t(32l);
+
+  let left_paren = t(40l);
+  let right_paren = t(41l);
+
+  let asterisk = t(42l);
+  let plus = t(43l);
+  let minus = t(45l);
+  let period = t(46l);
+  let slash = t(47l);
+  let caret = t(94l);
+
+  let equals = t(61l);
+
+  let digit0 = t(48l);
+  let digit1 = t(49l);
+  let digit2 = t(50l);
+  let digit3 = t(51l);
+  let digit4 = t(52l);
+  let digit5 = t(53l);
+  let digit6 = t(54l);
+  let digit7 = t(55l);
+  let digit8 = t(56l);
+  let digit9 = t(57l);
+
+  let pad_divide = t(1073741908l);
+  let pad_multiply = t(1073741909l);
+  let pad_minus = t(1073741910l);
+  let pad_plus = t(1073741911l);
+  let pad_period = t(1073741923l);
+
+  let pad_equals = t(1073741927l);
+
+  let p_digit1 = t(1073741913l);
+  let p_digit2 = t(1073741914l);
+  let p_digit3 = t(1073741915l);
+  let p_digit4 = t(1073741916l);
+  let p_digit5 = t(1073741917l);
+  let p_digit6 = t(1073741918l);
+  let p_digit7 = t(1073741919l);
+  let p_digit8 = t(1073741920l);
+  let p_digit9 = t(1073741921l);
+  let p_digit0 = t(1073741922l);
+
+  let c = t(99l);
+
+  let delete = t(127l);
+
+  let right = t(1073741903l);
+  let left = t(1073741904l);
 };
 
 module WheelType = {
