diff --git a/src/Makefile b/src/Makefile
index 92af4ce..9a86efe 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,3 +1,5 @@
+OCAMLFIND = ocamlfind
+
 all: opam-file-format.cmxa opam-file-format.cma
 
 %.ml: %.mll
@@ -7,13 +9,13 @@ all: opam-file-format.cmxa opam-file-format.cma
 	ocamlyacc $<
 
 %.cmi: %.mli
-	ocamlc -c $<
+	$(OCAMLFIND) ocamlc -c $<
 
 %.cmo: %.ml %.cmi
-	ocamlc -c $<
+	$(OCAMLFIND) ocamlc -c $<
 
 %.cmx: %.ml %.cmi
-	ocamlopt -c $<
+	$(OCAMLFIND) ocamlopt -c $<
 
 MODULES = $(sort $(basename $(wildcard *.ml *.mll *.mly)))
 
@@ -22,19 +24,19 @@ GENERATED = $(patsubst %.mly,%.ml,$(wildcard *.mly)) \
             $(patsubst %.mll,%.ml,$(wildcard *.mll))
 
 opam-file-format.cma: $(MODULES:%=%.cmo)
-	ocamlc -a $^ -o $@
+	$(OCAMLFIND) ocamlc -a $^ -o $@
 
 opam-file-format.cmxa: $(MODULES:%=%.cmx)
-	ocamlopt -a $^ -o $@
+	$(OCAMLFIND) ocamlopt -a $^ -o $@
 
 opam-file-format.cmxs: $(MODULES:%=%.cmx)
-	ocamlopt -shared $^ -o $@
+	$(OCAMLFIND) ocamlopt -shared $^ -o $@
 
 .PHONY: clean
 clean:
 	rm -f *.cm* *.o *.a $(GENERATED) .depend
 
 .depend: *.ml *.mli $(GENERATED)
-	ocamldep *.mli *.ml > .depend
+	$(OCAMLFIND) ocamldep *.mli *.ml > .depend
 
 -include .depend
