diff --git a/src/config/discover.re b/src/config/discover.re
index b743bb4..01f1331 100644
--- a/src/config/discover.re
+++ b/src/config/discover.re
@@ -1,40 +1,88 @@
-open Configurator.V1.Flags;
+open Configurator.V1;
 
 let getenv = name =>
-  try(Sys.getenv(name)) {
+  try (Sys.getenv(name)) {
   | Not_found => failwith("Error: Undefined environment variable: " ++ name)
   };
 
 type os =
-  | Windows
-  | Mac
+  | Android
+  | IOS
   | Linux
-  | Unknown;
-
-let uname = () => {
-  let ic = Unix.open_process_in("uname");
-  let uname = input_line(ic);
-  let () = close_in(ic);
-  uname;
-};
+  | Mac
+  | Windows;
 
-let get_os =
-  switch (Sys.os_type) {
-  | "Win32" => Windows
-  | _ =>
-    switch (uname()) {
-    | "Darwin" => Mac
-    | "Linux" => Linux
-    | _ => Unknown
-    }
+let header_detect_system = {|
+  #if __APPLE__
+    #include <TargetConditionals.h>
+    #if TARGET_OS_IPHONE
+      #define PLATFORM_NAME "ios"
+    #else
+      #define PLATFORM_NAME "mac"
+    #endif
+  #elif __linux__
+    #if __ANDROID__
+      #define PLATFORM_NAME "android"
+    #else
+      #define PLATFORM_NAME "linux"
+    #endif
+  #elif WIN32
+    #define PLATFORM_NAME "windows"
+  #endif
+|};
+let get_os = t => {
+  let header = {
+    let file = Filename.temp_file("discover", "os.h");
+    let fd = open_out(file);
+    output_string(fd, header_detect_system);
+    close_out(fd);
+    file;
+  };
+  let platform =
+    Configurator.V1.C_define.import(
+      t,
+      ~includes=[header],
+      [("PLATFORM_NAME", String)],
+    );
+  switch (platform) {
+  | [(_, String("android"))] => Android
+  | [(_, String("ios"))] => IOS
+  | [(_, String("linux"))] => Linux
+  | [(_, String("mac"))] => Mac
+  | [(_, String("windows"))] => Windows
+  | _ => failwith("Unknown operating system")
   };
+};
 
 let () = {
   let ccopt = s => ["-ccopt", s];
   let cclib = s => ["-cclib", s];
   let framework = s => ["-framework", s];
-  let flags =
-    switch (get_os) {
+  let flags = os =>
+    switch (os) {
+    | Android =>
+      []
+      @ ["-verbose"]
+      @ cclib("-lfreetype")
+      @ cclib("-lz")
+      @ cclib("-lskia")
+      @ cclib("-lSDL2")
+      @ cclib("-lGLESv2")
+      @ cclib("-lGLESv1_CM")
+      @ cclib("-lm")
+      @ cclib("-llog")
+      @ cclib("-landroid")
+      @ ccopt("-L" ++ getenv("FREETYPE2_LIB_PATH"))
+      @ ccopt("-L" ++ getenv("SDL2_LIB_PATH"))
+      @ ccopt("-L" ++ getenv("SKIA_LIB_PATH"))
+      @ ccopt("-L" ++ getenv("JPEG_LIB_PATH"))
+      @ ccopt("-I" ++ getenv("FREETYPE2_INCLUDE_PATH"))
+      @ ccopt("-I" ++ getenv("SKIA_INCLUDE_PATH"))
+      @ cclib("-ljpeg")
+      @ ccopt("-I/usr/include")
+      @ ccopt("-lstdc++")
+    | IOS
+    | Mac => []
     | Linux =>
       []
       @ ["-verbose"]
@@ -53,25 +101,39 @@ let () = {
       @ cclib("-ljpeg")
       @ ccopt("-I/usr/include")
       @ ccopt("-lstdc++")
-
     | Windows =>
       []
       @ cclib("-lskia")
       @ cclib("-lSDL2")
       @ ccopt("-L" ++ getenv("SDL2_LIB_PATH"))
       @ ccopt("-L" ++ getenv("SKIA_LIB_PATH"))
-
-    | _ => []
     };
 
-  let cflags =
-    switch (get_os) {
+  let cflags = os =>
+    switch (os) {
+    | Android =>
+      []
+      @ ["-lSDL2"]
+      @ ["-lGLESv2"]
+      @ ["-lGLESv1_CM"]
+      @ ["-lm"]
+      @ ["-llog"]
+      @ ["-landroid"]
+      @ ["-lskia"]
+      @ ["-I" ++ getenv("SDL2_INCLUDE_PATH")]
+      @ ["-I" ++ getenv("SKIA_INCLUDE_PATH")]
+      @ ["-I" ++ getenv("SKIA_INCLUDE_PATH") ++ "/c"]
+      @ ["-L" ++ getenv("SKIA_LIB_PATH")]
+      @ ["-L" ++ getenv("SDL2_LIB_PATH")]
+      @ ["-L" ++ getenv("JPEG_LIB_PATH")]
+      @ ["-lstdc++"]
+      @ ["-ljpeg"]
+    | IOS
     | Mac =>
       []
       @ ["-I" ++ getenv("SDL2_INCLUDE_PATH")]
       @ ["-I" ++ getenv("SKIA_INCLUDE_PATH")]
       @ ["-I" ++ getenv("SKIA_INCLUDE_PATH") ++ "/c"]
-
     | Linux =>
       []
       @ ["-lSDL2"]
@@ -84,19 +146,63 @@ let () = {
       @ ["-L" ++ getenv("JPEG_LIB_PATH")]
       @ ["-lstdc++"]
       @ ["-ljpeg"]
-
     | Windows =>
       []
       @ ["-std=c++1y"]
       @ ["-I" ++ getenv("SDL2_INCLUDE_PATH")]
       @ ["-I" ++ getenv("SKIA_INCLUDE_PATH")]
       @ ["-I" ++ getenv("SKIA_INCLUDE_PATH") ++ "/c"]
-
-    | _ => failwith("cflags: unknown platform")
     };
 
-  let libs =
-    switch (get_os) {
+  let libs = os =>
+    switch (os) {
+    | Android =>
+      []
+      @ [
+        "-lSDL2",
+        "-lGLESv2",
+        "-lGLESv1_CM",
+        "-lm",
+        "-llog",
+        "-landroid",
+        "-lskia",
+        "-lfreetype",
+        "-lz",
+        "-L" ++ getenv("JPEG_LIB_PATH"),
+        "-ljpeg",
+        "-lstdc++",
+        "-L" ++ getenv("SDL2_LIB_PATH"),
+        "-L" ++ getenv("SKIA_LIB_PATH"),
+        "-L" ++ getenv("FREETYPE2_LIB_PATH"),
+      ]
+    | IOS =>
+      []
+      @ ["-L" ++ getenv("JPEG_LIB_PATH")]
+      @ ["-L" ++ getenv("SKIA_LIB_PATH")]
+      @ ["-L" ++ getenv("FREETYPE2_LIB_PATH")]
+      @ ["-L" ++ getenv("SDL2_LIB_PATH")]
+      @ framework("OpenGLES")
+      @ framework("UIKit")
+      @ framework("Foundation")
+      @ framework("GameController")
+      @ framework("AVFoundation")
+      @ framework("QuartzCore")
+      @ framework("CoreMotion")
+      @ framework("CoreFoundation")
+      @ framework("CoreAudio")
+      @ framework("CoreVideo")
+      @ framework("CoreServices")
+      @ framework("CoreGraphics")
+      @ framework("CoreText")
+      @ framework("CoreFoundation")
+      @ framework("AudioToolbox")
+      @ framework("IOKit")
+      @ framework("Metal")
+      @ ["-liconv"]
+      @ ["-lSDL2"]
+      @ ["-lskia"]
+      @ ["-lstdc++"]
+      @ [getenv("JPEG_LIB_PATH") ++ "/libturbojpeg.a"]
     | Mac =>
       []
       @ ["-L" ++ getenv("JPEG_LIB_PATH")]
@@ -121,7 +227,6 @@ let () = {
       @ ["-lskia"]
       @ ["-lstdc++"]
       @ [getenv("JPEG_LIB_PATH") ++ "/libturbojpeg.a"]
-
     | Linux =>
       []
       @ [
@@ -139,7 +244,6 @@ let () = {
         "-L" ++ getenv("SKIA_LIB_PATH"),
         "-L" ++ getenv("FREETYPE2_LIB_PATH"),
       ]
-
     | Windows =>
       []
       @ ["-luser32"]
@@ -148,17 +252,18 @@ let () = {
       @ ["-lstdc++"]
       @ ["-L" ++ getenv("SDL2_LIB_PATH")]
       @ ["-L" ++ getenv("SKIA_LIB_PATH")]
-
-    | _ => failwith("libs: Unknown platform")
     };
 
-  write_sexp("flags.sexp", flags);
-  write_lines("c_flags.txt", cflags);
-  write_sexp("c_flags.sexp", cflags);
-  write_sexp("c_library_flags.sexp", libs);
-  write_lines("c_library_flags.txt", libs);
-  write_sexp(
-    "cclib_c_library_flags.sexp",
-    libs |> List.map(o => ["-cclib", o]) |> List.flatten,
-  );
+  main(~name="c_test", t => {
+    let os = get_os(t);
+    Flags.write_sexp("flags.sexp", flags(os));
+    Flags.write_lines("c_flags.txt", cflags(os));
+    Flags.write_sexp("c_flags.sexp", cflags(os));
+    Flags.write_sexp("c_library_flags.sexp", libs(os));
+    Flags.write_lines("c_library_flags.txt", libs(os));
+    Flags.write_sexp(
+      "cclib_c_library_flags.sexp",
+      libs(os) |> List.map(o => ["-cclib", o]) |> List.flatten,
+    );
+  });
 };
